lib/Extractor/KLEEBuilder.cpp

```diff
@@ -169,6 +172,21 @@ ref<Expr> ExprBuilder::makeSizedArrayRead(unsigned Width, StringRef Name,
                                                                                   klee::ConstantExpr::create(i, Width))));
       PowerTwoBitsMap[Origin] = PowerExpr;
     }
+    if (Origin->NumSignBits > 1) {
+      ref<Expr> SignBitsExpr = klee::ConstantExpr::create(1, Expr::Bool);
+      ref<Expr> Res = LShrExpr::create(Var, klee::ConstantExpr::create(Width-Origin->NumSignBits, Width));
+      ref<Expr> TestOnes = SubExpr::create(ShlExpr::create(klee::ConstantExpr::create(1, Width),
+                                           klee::ConstantExpr::create(Origin->NumSignBits, Width)),
+                                           klee::ConstantExpr::create(1, Width));
+      SignBitsExpr = AndExpr::create(SignBitsExpr,
+                                     OrExpr::create(EqExpr::create(Res, TestOnes),
+                                                    EqExpr::create(Res, klee::ConstantExpr::create(0, Width))));
+      ref<Expr> MSB = ExtractExpr::create(Var, Width-1, Expr::Bool);
+      SignBitsExpr = AndExpr::create(SignBitsExpr,
+                                     Expr::createImplies(EqExpr::create(MSB, klee::ConstantExpr::create(0, Expr::Bool)),
+                                                         EqExpr::create(ctlz(Var), klee::ConstantExpr::create(Origin->NumSignBits, Width))));
```

Comment:
I'm pretty sure that arithmetic shift right followed by assertion that the result == 0 or result == -1 is the simplest and best implementation for this